Revision: 34313

JoaquinHenley opened the pull request:
【内置sql】使用内置SQL插入timestamp类型记录，1970年前的时间戳记录异常

【重现步骤】 
db.execUpdate( 'insert into cs.cl( age, name, birthday ) values( 25, "Tom", timestamp("1969-01-01-00:00:00.000000"))' ) 
Takes 0.000948s. 
【预期结果】 
查询出1969年的时间戳 
【实际结果】 
db.cs.cl.find() 
{ 
  "_id": { 
    "$oid": "5a3b7dc5c2f011b9095ecac2" 
        }, 
  "age": 25, 
  "name": "Tom", 
  "birthday": { 
    "$timestamp": "2009-04-15-05.45.19.000000" 
        } 
} 
Return 1 row(s).

DebbieMadisonS commented:
请安排分析

PhillipsPaull commented:
请处理

JoaquinHenley commented:
【原因分析】 
内置sql插入时间戳时调用c++驱动的appendTimestamp接口 
BSONObjBuilder& BSONObjBuilder::appendTimestamp(const StringData& fieldName, unsigned long long time, unsigned int inc) 
{ 
   …… 
   xxx = time/1000; => time传入是一个毫秒数，真正存储是存秒数，所以要除以1000。 
   …… 
} 
1970年以前绝对秒数为负数，time实际上是个负值，time/1000时没有把time转成long long 处理。

JoaquinHenley commented:
【自验步骤】 
1、通过内置sql验证cpp驱动的appendTimestamp 
db.execUpdate( 'insert into cs.cl( a ) values( timestamp("1902-01-01-01:00:00.000000") )' ) 
db.exec( 'select * from cs.cl' ) 
{ 
  "_id": { 
    "$oid": "5a812daab525e3d733ca2eb1" 
  }, 
  "a": { 
    "$timestamp": "1902-01-01-00.00.00.000000" 
  } 
} 

2、验证cpp驱动的Date_t.toString() 
time_t s = -1325494801 ; =>1902-01-01 00:00:00 
long long millis = s*1000 ; 
Date_t date ( millis ) ; 
cout<<"date is: "<<date.toString ()<<endl ; 
结果：date is: Wed Jan 1 00:05:52 1902

JoaquinHenley commented:
提交主干 
Revision: 34313 

Date: 2018年2月13日 13:56:57 
Message: 
3133 使用内置SQL插入timestamp类型记录，1970年前的时间戳记录异常 

Modified : /trunk/SequoiaDB/engine/bson/bson_db.h 
Modified : /trunk/SequoiaDB/engine/bson/bsonobjbuilder.h 
Modified : /trunk/SequoiaDB/engine/bson/util/misc.h 
Modified : /trunk/SequoiaDB/engine/bson/util/optime.h 
Modified : /trunk/SequoiaDB/engine/qgm/qgmUtil.cpp

JoaquinHenley commented:
提交分支2.6 
Revision: 34358 

Date: 2018年2月23日 11:02:03 
Message: 
3133 使用内置SQL插入timestamp类型记录，1970年前的时间戳记录异常 

Modified : /branches/sequoiadb_2.6/SequoiaDB/engine/bson/bson_db.h 
Modified : /branches/sequoiadb_2.6/SequoiaDB/engine/bson/bsonobjbuilder.h 
Modified : /branches/sequoiadb_2.6/SequoiaDB/engine/bson/util/misc.h 
Modified : /branches/sequoiadb_2.6/SequoiaDB/engine/bson/util/optime.h

JoaquinHenley commented:
【检查项】 
检查项---------------------------------------------------------结论（NA/N/Y） 

修改代码是否全部提交SVN主干和分支版本 
且符合代码规范-------------------------------------------- Y（主干 分支2.8 分支2.6）备注：分支2.6 内置SQL中没有timestamp功能 
是否用例全部通过----------------------------------------- Y 
是否清淅标明重现步骤----------------------------------- Y 
修改代码是否通过检视----------------------------------- N 
资料是否变更---------------------------------------------- Y （c++ bson接口资料变更） 
接口是否变更---------------------------------------------- Y （接口从unsigned变成signed，但是不影响兼容性） 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容---------------------------------------- Y 
是否影响安装部署和监控-------------------------------- N 
是否对影响性能------------------------------------------- N 
sdb shell帮助信息是否更新----------------------------- NA

JoaquinHenley commented:
修改主干资料： 
Revision: 34533 

Date: 2018年2月28日 16:53:20 
Message: 
3133 增加资料中对appendTimestamp/OpTime/Date_t的描述 

Modified : /trunk/SequoiaDB/engine/bson/bsonobjbuilder.h 
Modified : /trunk/doc/config/doxygen/doxygen-bsoncpp.conf

JoaquinHenley commented:
修改2.8资料：svn号34535 
修改2.6资料：svn号34538

JoaquinHenley commented:
请更新2.6的bsoncpp资料

AmyChristopherD commented:
2.6的bsoncpp资料更新完成。

JoaquinHenley commented:
已处理

PhillipsPaull commented:
请安排测试，主要关注兼容性。

DebbieMadisonS commented:
请验证

BettyFrancesZ commented:
请验证

FairyCook commented:
【验证结果】：验证不通过 
【回归版本】： 
 SequoiaDB version: 2.9 
Release: 34625 
2018-03-07-00.38.02(Enterprise-Debug) 
【验证步骤】 
1、构造Date_t数据，指定日期为0000-01-01的毫秒值，使用toString方法输出后，结果多了一个空白行 
SINT64 mills = -62167248352000 ; // 0000-01-01 00:00:00 
Date_t dt( mills ) ; 
cout << dt.toString() << endl ;

DebbieMadisonS commented:
请安排分析

PhillipsPaull commented:
请重新处理

JoaquinHenley commented:
【原因分析】 
年份小于1000的日期，toString()后都会多出一个空白行。 
因为为了去掉最后一个换行，它指定第24个字符为结束符。但是年份小于1000的第23个字符就是换行了。 

【修改描述】 
用strlen取字符串的长度，然后再置最后一个字符为结束符。

JoaquinHenley commented:
提交2.6 
Revision: 34990 

Date: 2018年3月26日 9:44:39 
Message: 
3133 10000年以前 Date_t.toString()返回的字符串多了一个换行 

Modified : /branches/sequoiadb_2.6/SequoiaDB/engine/bson/util/misc.h 

提交2.8 
Revision: 34989 

Date: 2018年3月26日 9:42:39 
Message: 
3133 10000年以前 Date_t.toString()返回的字符串多了一个换行 

Modified : /branches/sequoiadb_2.8/SequoiaDB/engine/bson/util/misc.h 

提交主干 
Revision: 34988 

Date: 2018年3月26日 9:42:01 
Message: 
3133 10000年以前 Date_t.toString()返回的字符串多了一个换行 

Modified : /trunk/SequoiaDB/engine/bson/util/misc.h

JoaquinHenley commented:
【自验】 
1000年以前的日期，toString()结果没有换行 

【检查项】 
检查项---------------------------------------------------------结论（NA/N/Y） 
-------------------------------------------------------------------------------------------- 
修改代码是否全部提交SVN主干和分支版本 
且符合代码规范-------------------------------------------- Y（主干 分支2.8 分支2.6） 
是否用例全部通过----------------------------------------- Y 
是否清淅标明重现步骤----------------------------------- Y 
修改代码是否通过检视----------------------------------- N 
资料是否变更---------------------------------------------- Y （c++ bson接口资料变更） 
接口是否变更---------------------------------------------- Y （接口从unsigned变成signed，但是不影响兼容性） 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容---------------------------------------- Y 
是否影响安装部署和监控-------------------------------- N 
是否对影响性能------------------------------------------- N 
sdb shell帮助信息是否更新----------------------------- NA

PhillipsPaull commented:
请安排测试

DebbieMadisonS commented:
请验证

BettyFrancesZ commented:
请验证

FairyCook commented:
【验证结果】：验证问题未解决 
【回归版本】： 
2.9 
SequoiaDB version: 2.9 
Release: 35103 
2018-04-03-00.39.56(Enterprise-Debug) 
【验证步骤】 
1、验证Date_t( long long m )，设置毫秒数为日期有效取值范围0000-01-01到9999-12-31的毫秒数，使用Date_t.toString输出结果正常 
2、验证appendTimestamp( const StringData& fieldName )，构造bson对象后输出结果正常 
3、验证OpTime( signed a, unsigned b )及appendTimestamp( const StringData& fieldName, long long val )，设置OpTime中的毫秒数和微秒数为时间戳有效取值范围1902-01-01 00:00:00.000000至2037-12-31 23.59.59:999999，构造bson对象输出后结果正确 
4、验证appendTimestamp( const StringData& fieldName, long long time, unsigned int inc )，设置毫秒数和微秒数为时间戳有效取值范围1902-01-01 00:00:00.000000至2037-12-31 23.59.59:999999，构造bson对象输出后结果正确 
5、验证兼容性，使用旧版驱动执行以上操作，结果正常 
6、验证资料，检查驱动bson文档，文档内容存在错误，错误信息参见附件 
【测试用例】 
TestLink： 
seqDB-14702:接口Date_t.toString()测试 
seqDB-14704:接口BSONObjBuilder::appendTimestamp校验 
SVN： 
trunk\testcase_new\driver\cpp\date\dateTime12219_12220.cpp 
【备注】： 
部分接口的参数名称太简单，如Date_t( long long m )可以修改为Date_t( long long millis )，OpTime( signed a, unsigned b )可以修改为OpTime( signed secs, unsigned microsecs )，其他接口存在类似情况

JoaquinHenley commented:
针对提的问题： 
1、资料的错误已修改 
2、cpp驱动用的是mongodb的bson库，对接口要求没必要那么严格

DebbieMadisonS commented:
请验证

BettyFrancesZ commented:
【验证结果】：验证问题已经解决，请检视 

【回归版本】 
3.0 
【验证步骤】 
文档-》驱动-》C++驱动-》C++ BSON使用-》C++ BSON API -》 BSONObjBuilder& bson::BSONObjBuilder::appendTimestamp() 
Timestamp is a special BSON datatype. 
Append a timestamp element to the object being built. 
描述已修改，描述正确 
【测试用例】 
不需要补充用例

AmyChristopherD commented:
开发有提交代码到2.6和2.8版本，请补充在2.6和2.8版本上验证

BettyFrancesZ commented:
【补充验证】 
验证通过，请检视 
2.8、2.6版本驱动文档已修复描述，与3.0版本一致

AustinAngusE commented:
需要在2.6和2.8版本上验证功能

BettyFrancesZ commented:
请补充验证2.6/2.8驱动功能

EthanJames12 commented:
【验证结果】 
验证问题已经解决，请检视 
【回归版本】 
SequoiaDB shell version: 2.8.5 
Release: 36863 
2018-09-19-17.30.08(Enterprise-Debug) 

SequoiaDB shell version: 2.6.5 
Release: 36868 
2018-09-18-23.41.54(Enterprise-Debug) 
【验证步骤】 
1、在v2.6和2.8的版本上执行可执行程序trunk\testcase_new\driver\cpp\date\dateTime12219_12220.cpp，驱动功能均无异常 
  build/normal/dateTime12219_12220 -n 192.168.31.57 -s 11810 
HOSTNAME : 192.168.31.57 
SVCNAME : 11810 
CHANGEDPREFIX : sdv_cpp_test 
RSPVPORTBEGIN : 26000 
RSPVPORTEND : 27000 
RSPVNODEDIR : /opt/sequoiadb/database/ 
WORKDIR : /tmp/cpptest 
COORDURL : 192.168.31.57:11810 
FORCECLEAR : 1 
[==========] Running 5 tests from 5 test cases. 
[----------] Global test environment set-up. 
[----------] 1 test from dateTest12219 
[ RUN ] dateTest12219.appendDate 
[ OK ] dateTest12219.appendDate (0 ms) 
[----------] 1 test from dateTest12219 (0 ms total) 

[----------] 1 test from timestampTest12220 
[ RUN ] timestampTest12220.appendTimestamp 
[ OK ] timestampTest12220.appendTimestamp (0 ms) 
[----------] 1 test from timestampTest12220 (0 ms total) 

[----------] 1 test from dateTest14702 
[ RUN ] dateTest14702.toString 
[ OK ] dateTest14702.toString (0 ms) 
[----------] 1 test from dateTest14702 (0 ms total) 

[----------] 1 test from dateTest14703 
[ RUN ] dateTest14703.append 
[ OK ] dateTest14703.append (0 ms) 
[----------] 1 test from dateTest14703 (0 ms total) 

[----------] 1 test from timestampTest14704 
[ RUN ] timestampTest14704.appendTimestamp 
[ OK ] timestampTest14704.appendTimestamp (0 ms) 
[----------] 1 test from timestampTest14704 (0 ms total) 

[----------] Global test environment tear-down 
[==========] 5 tests from 5 test cases ran. (6 ms total) 
[ PASSED ] 5 tests. 
[root@r520-2 cpp]# pwd 
/data/disk3/sequoiadb/trunk/testcase_new/driver/cpp 
[root@r520-2 cpp]# build/normal/dateTime12219_12220 -n 192.168.31.57 -s 11810 
HOSTNAME : 192.168.31.57 
SVCNAME : 11810 
CHANGEDPREFIX : sdv_cpp_test 
RSPVPORTBEGIN : 26000 
RSPVPORTEND : 27000 
RSPVNODEDIR : /opt/sequoiadb/database/ 
WORKDIR : /tmp/cpptest 
COORDURL : 192.168.31.57:11810 
FORCECLEAR : 1 
[==========] Running 5 tests from 5 test cases. 
[----------] Global test environment set-up. 
[----------] 1 test from dateTest12219 
[ RUN ] dateTest12219.appendDate 
[ OK ] dateTest12219.appendDate (0 ms) 
[----------] 1 test from dateTest12219 (0 ms total) 

[----------] 1 test from timestampTest12220 
[ RUN ] timestampTest12220.appendTimestamp 
[ OK ] timestampTest12220.appendTimestamp (0 ms) 
[----------] 1 test from timestampTest12220 (0 ms total) 

[----------] 1 test from dateTest14702 
[ RUN ] dateTest14702.toString 
[ OK ] dateTest14702.toString (0 ms) 
[----------] 1 test from dateTest14702 (0 ms total) 

[----------] 1 test from dateTest14703 
[ RUN ] dateTest14703.append 
[ OK ] dateTest14703.append (0 ms) 
[----------] 1 test from dateTest14703 (0 ms total) 

[----------] 1 test from timestampTest14704 
[ RUN ] timestampTest14704.appendTimestamp 
[ OK ] timestampTest14704.appendTimestamp (0 ms) 
[----------] 1 test from timestampTest14704 (0 ms total) 

[----------] Global test environment tear-down 
[==========] 5 tests from 5 test cases ran. (0 ms total) 
[ PASSED ] 5 tests. 

[root@r520-2 cpp]# build/normal/dateTime12219_12220 -n 192.168.31.55 -s 11810 
HOSTNAME : 192.168.31.55 
SVCNAME : 11810 
CHANGEDPREFIX : sdv_cpp_test 
RSPVPORTBEGIN : 26000 
RSPVPORTEND : 27000 
RSPVNODEDIR : /opt/sequoiadb/database/ 
WORKDIR : /tmp/cpptest 
COORDURL : 192.168.31.55:11810 
FORCECLEAR : 1 
[==========] Running 2 tests from 2 test cases. 
[----------] Global test environment set-up. 
[----------] 1 test from dateTest12219 
[ RUN ] dateTest12219.date12219 
[ OK ] dateTest12219.date12219 (0 ms) 
[----------] 1 test from dateTest12219 (0 ms total) 

[----------] 1 test from timestampTest12220 
[ RUN ] timestampTest12220.timestamp12220 
[ OK ] timestampTest12220.timestamp12220 (0 ms) 
[----------] 1 test from timestampTest12220 (0 ms total) 

[----------] Global test environment tear-down 
[==========] 2 tests from 2 test cases ran. (0 ms total) 
[ PASSED ] 2 tests. 
You have mail in /var/spool/mail/root 

【测试用例】 
不需要补充用例
